packages/react-router/modules/__tests__/Switch-test.js

```diff
@@ -282,6 +282,67 @@ describe("A <Switch>", () => {
     expect(node.innerHTML).toMatch(/one/);
   });
 
+  it("does handle fragments", () => {
+    const node = document.createElement("div");
+
+    ReactDOM.render(
+      <MemoryRouter initialEntries={["/two"]}>
+        <Switch>
+          <Route path="/one" render={() => <h1>one</h1>} />
+          <Fragment>
+            <Route path="/two" render={() => <h1>two</h1>} />
+          </Fragment>
+          <Route path="/three" render={() => <h1>three</h1>} />
+        </Switch>
+      </MemoryRouter>,
+      node
+    );
+
+    expect(node.innerHTML).toMatch(/two/);
+  });
+
+  it("does handle nested fragments", () => {
```

Comment:
We specifically should not support this. We don't support nested routes in a Switch, so this behavior difference will undoubtedly cause confusion. To be clear, I see this:
```js

<Switch>
  <Route path="/foo" component={Foo} />
  <>
    <Route path="/bar" component={Bar} />
    <Route path="/baz" component={Baz} />
  </>
</Switch>
```
as equivalent to this:
```js
<Switch>
  <Route path="/foo" component={Foo} />
  {[barRoute, bazRoute].map(route => <Route path={route.path} component={route.component} />)}
</Switch>
````
Fragments are basically convenience functions for arrays. So, I'd treat them as such to maintain the status quo with Switch's behaviors.

Reply:
OK, I'm convinced. I didn't know about that behavior. Thanks for explaining!

Reply:
Actually, I'm not. Sorry, just busted out some code here and contradicted myself. Apologies for the flip flop!
```js
React.Children.count([<div/>, <div/>, [<div />, <div/>]]) == 4
React.Children.count([
  <div/>,
  <div/>,
  <React.Fragment>
    <div/>
    <div/>
  </React.Fragment>
]) == 3
```
Fragments are not the same as arrays, they just have a similar use case. The post here is clarifying: https://github.com/facebook/react/issues/11859#issuecomment-355654128

I'd much rather do `React.Children.map(children, child => child.type === React.Fragment ? child.props.children : child).forEach(child => ...)`. It's not common to have nested child arrays inside of `<Switch>`, so trying to emulate that behavior with something that's intentionally inconsistent.

Reply:
I do not see why fragments should be handled differently for all intents and purposes of react-router than arrays. They are element containers and should be handled accordingly. 

Design decisions have been made for fragments which result in different behavior for `React.Children.count`, `React.Children.map` and others when comparing fragment and array behavior. This difference in behavior stems from important differences when handling these containers as part of `render`. React-router doesn't have to behave different when encountering fragments vs. arrays.

Furthermore, it makes no sense for react-router to try to `matchPath` a fragment element. This will never work.

Lastly, there is no technical reason why fragment recursion cannot be supported. There is no additional cost associated to it. It is faster than a `map` followed by a `forEach`. It makes array/fragment behavior consistent.

Reply:
> I do not see why fragments should be handled differently for all intents and purposes of react-router than arrays. 

Because they're not arrays. I'd really rather not try to go against the upstream behavior semantics, even if it is admittedly weird. We got burned in previous versions by trying to do too much of React's behavior on our own. The edge cases are death by a thousand cuts.

I'm still on Team No Flatten™, but the rules dictate we have two maintainers approve before merging. If anyone else wants to voice their opinions, even if it's different from my own, they can and we can work towards a resolution.

Reply:
I'm not a maintainer, but I recommended this change over in #5785 , so I thought I'd voice that I'm also on Team No Flatten for whatever that's worth. Thanks for all of your work @timdorr and @bripkens!

Reply:
Added the requested example to the issue:

https://github.com/ReactTraining/react-router/issues/5785#issuecomment-361852111